Add Video Swin Transformer

innat opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons --sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes https://github.com/keras-team/keras-cv/issues/2262


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

innat commented:
@divyashreepathihalli @tirthasheshpatel 
This PR is ready to take some initial feedback. 

I have some query. 
- What does `kaggle_handle` mean? (A weight file or folder! If so, how [this config](https://github.com/keras-team/keras-cv/blob/9dd547a725c081db0264023717f43d14301ea3d2/keras_cv/models/backbones/efficientnet_v2/efficientnet_v2_backbone_presets.py#L17-L27) contains weight path!). How can I test the model with local path?  Setting `"kaggle_handle": /usr/video_swin_weights"` didn't work for me.
- How should I hand over the weight files, (which were ported from official release)? Should it be `.weight.h5` or `.keras` format?

tirthasheshpatel commented:
Thanks for the PR @innat! This is really great work!

> * What does `kaggle_handle` mean? (A weight file or folder! If so, how [this config](https://github.com/keras-team/keras-cv/blob/9dd547a725c081db0264023717f43d14301ea3d2/keras_cv/models/backbones/efficientnet_v2/efficientnet_v2_backbone_presets.py#L17-L27) contains weight path!). How can I test the model with local path?  Setting `"kaggle_handle": /usr/video_swin_weights"` didn't work for me.

`kaggle_handle` is the link to the Kaggle model card. It contains the weights in `.weights.h5` format and the config for the pretrained model as `config.json`. For example, the link `kaggle://keras/efficientnetv2/keras/efficientnetv2_s/2` points to [this Kaggle model card](https://www.kaggle.com/models/keras/efficientnetv2).

To test the model locally, I usually just initialize the pretrained model and call `load_weights` to load the pretrained weights. You can do this in tests to make sure they pass. Once you have the weights ready, feel free to ping me, @divyashreepathihalli, or @mattdangerw and we can get the weights uploaded to Kaggle.

> * How should I hand over the weight files, (which were ported from official release)? Should it be `.weight.h5` or `.keras` format?

It should be a `.weights.h5` file and can be generated using the `.save_weights` method.

innat commented:
@tirthasheshpatel Thanks for the details response. About weight and keras version, one thing I like to share, which I also reported [here](https://github.com/keras-team/keras/issues/19238). Sharing here, just in case anyone faces the same, or  in case I've been inadvertently overlooked something.

---

If we use `keras==3.0.4` (or < `3.0.5`), the logits scores between `torchvision.swin3d` and `keras_cv.video_swin` (with `torch` backend) get matched. But not with `tf` and `jax` backend but these two backend scores match to each other. Now, from `keras==3.0.5` (or `nightly`), all bakends give similar scores and wrong compared to `torchvision.swin3d`.

[colab.](https://colab.research.google.com/drive/1fKPp7NYctgSZKrkkQUCbq4KwHxIV0Z5W?usp=sharing)

```python
keras==3.0.4
torch_out = torchvision.swin3d

keras_tf_out = keras_cv.video_swin # tf backend
keras_jax_out = keras_cv.video_swin # jax backend
keras_torch_out = keras_cv.video_swin # torch backend

torch_out == keras_torch_out 
torch_out != [keras_tf_out, keras_jax_out]
keras_tf_out == keras_jax_out
```

[colab.](https://colab.research.google.com/drive/1XEeKC8rhPAr-zglwI1kmUabpMIL4cH7B?usp=sharing)
```python
keras-3.0.5
torch_out = torchvision.swin3d

keras_tf_out = keras_cv.video_swin # tf backend
keras_jax_out = keras_cv.video_swin # jax backend
keras_torch_out = keras_cv.video_swin # torch backend

torch_out != keras_torch_out 
torch_out != [keras_tf_out, keras_jax_out]
keras_tf_out == keras_jax_out == keras_torch_out 
```

innat commented:
> Once you have the weights ready, feel free to ping me, @divyashreepathihalli, or @mattdangerw and we can get the weights uploaded to Kaggle.

@mattdangerw 
Here are the weights files, for three variation of swin model. 
- [x] [vidoe-swin-tiny](https://drive.google.com/file/d/1fWDSts4CiWT8s7OsrNhSzGlUgxRDKqix/view?usp=sharing)
- [x] [video-swin-small](https://drive.google.com/file/d/1og7LOMSe549YWPIhnKhAuB95Fli16PFT/view?usp=sharing)
- [x] [video-swin-base](https://drive.google.com/file/d/1oK9xBU2QQI8VWUPxXBnNOK3UD2vfDMQ9/view?usp=sharing)

To compare with official torch model, check this [gist](https://www.kaggle.com/code/ipythonx/video-swin-transformer-kerascv-and-torchvision).

innat commented:
@keras-team
A gentle reminder for this PR. 
I would like to complete it within this month (if possible) after addressing the reviews. Otherwise please let me know if there are some other high priority, so I will close this PR for the time being. Thanks.

divyashreepathihalli commented:
LGTM! Thanks for the PR @innat, can you also please provide an example use code snippet for Kaggle model page?

Also if you want to test GPU tests locally, you can run
`pytest <file name> --run_large`

innat commented:
@divyashreepathihalli 
Thanks. 

I've written plenty of eample how to use this model with different backend, [link](https://github.com/innat/VideoSwin/tree/main/guides). For kaggle model page, could you please specify what and how much demonstration is needed?

I've tested with GPU. It seems all are set. 

```python
!pytest .../keras_cv/layers/video_swin_layers_test.py --run_large
============================= test session starts ==============================
platform linux -- Python 3.10.13, pytest-8.0.1, pluggy-1.3.0
rootdir: /kaggle/working/keras-cv
configfile: setup.cfg
plugins: anyio-4.2.0, typeguard-4.1.5
collected 9 items                                                              

keras_cv/layers/video_swin_layers_test.py s..s.....                      [100%]

========================= 7 passed, 2 skipped in 1.52s =========================

!pytest ...//keras-cv/keras_cv/models/backbones/video_swin/ --run_large
============================= test session starts ==============================
platform linux -- Python 3.10.13, pytest-8.0.1, pluggy-1.3.0
rootdir: /kaggle/working/keras-cv
configfile: setup.cfg
plugins: anyio-4.2.0, typeguard-4.1.5
collected 16 items                                                             

keras_cv/models/backbones/video_swin/video_swin_backbone_presets_test.py s [  6%]
.sss..ss                                                                 [ 56%]
keras_cv/models/backbones/video_swin/video_swin_backbone_test.py s.sssss [100%]

======================== 4 passed, 12 skipped in 6.30s =========================


!pytest .../keras-cv/keras_cv/models/classification/video_classifier_test.py --run_large
============================= test session starts ==============================
platform linux -- Python 3.10.13, pytest-8.0.1, pluggy-1.3.0
rootdir: /kaggle/working/keras-cv
configfile: setup.cfg
plugins: anyio-4.2.0, typeguard-4.1.5
collected 8 items                                                              

keras_cv/models/classification/video_classifier_test.py s.......         [100%]

=================== 7 passed, 1 skipped in 221.61s (0:03:41) ===================
add Codeadd Markdown
````

divyashreepathihalli reviewed (approved):
Thanks for the PR! LGTM!

divyashreepathihalli reviewed (changes requested):
Hi @innat, please run `pytest  --run_large` on the everything! looks like the GPU tests are failing.

innat commented:
@divyashreepathihalli 
Thanks for checking. Sorry for the inconveniences. I only checked with the tensorflow backend. I break down the key error. Let me know if I've missed anyting.

- tensorflow - test passed - failling for segformer model
- torch - segmentation fault
- jax - test failing - keras.ops.ccond and system OOM

The jax issue is failing for [this](https://github.com/keras-team/keras/issues/19379), I think it's kinda bug in jax backend.

innat commented:
About weight, please check [this](https://github.com/keras-team/keras-cv/pull/2369#issuecomment-1989420576). 

weight identifier

```bash
a. kinetics400-tiny
b. kinetics400-samll
c. kinetics400-base
d. kinetics400-base-imgnet22k
f. kinetics600-base-imgnet22k
g. something-something-v2-base
```

Each has its backbone and classifier part.

comparision.

I showed comparision between kcv and torchvision, [gist](https://github.com/innat/VideoSwin/blob/main/guides/video-swin-transformer-kerascv-and-torchvision.ipynb). But note, torchvision offeres only a, b, c, d in the above weight identifier. So, in that gist, only these are shown.

innat commented:
I think test are all passed. But eventually failling for this https://github.com/keras-team/keras-cv/pull/2401

tirthasheshpatel reviewed (commented):
Just wanted to chime in and say that this is a very high quality contribution, thanks @innat, great work! Will take a deeper look later today.

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+@keras_cv_export(
+    "keras_cv.layers.VideoSwinPatchingAndEmbedding", package="keras_cv.layers"
+)
+class VideoSwinPatchingAndEmbedding(keras.Model):
```
```suggestion
class VideoSwinPatchingAndEmbedding(keras.Layer):
```

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+        return config
+
+
+@keras_cv_export(
+    "keras_cv.layers.VideoSwinPatchingAndEmbedding", package="keras_cv.layers"
+)
```
I don't think we should export this. It's not present in the `keras_cv.layers` public API so exporting it has no effect.

innat replied:
This is already present in the `keras_cv.layers` ` __init__.py` https://github.com/keras-team/keras-cv/pull/2369#discussion_r1545523595

tirthasheshpatel replied:
Yeah, I meant not adding this in the public API unless useful output of Video Swin Transformer architecture.

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+# limitations under the License.
+
+import numpy as np
+from keras import layers
```
Let's remove this import and use `keras.layers` instead. This is a good pattern to follow until Keras 2 support is dropped. We can then get rid of the `from keras_cv.backend import keras` and do keras imports directly.

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+        return config
+
+
+class VideoSwinWindowAttention(keras.Model):
```
```suggestion
class VideoSwinWindowAttention(keras.Layer):
```

innat replied:
Subclassing to `keras.Model` instead of layer is on purpose. 

Right now, because of using `keras.Model`,  if we use `model.summary(expand_nested=True)`, it will disclose the sub-layers summary as expected. But not with `keras.Layer`. 

With `keras.Model`

```python
with_model_swin.summary(expand_nested=True)
Model: "new_swin_transformer3d"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 8, 256, 256, 3) │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_patching_and_embeddi… │ (None, 4, 64, 64, 96)  │         9,504 │
│ (VideoSwinPatchingAndEmbedding) │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ embed_proj (Conv3D)        │ ?                      │         9,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ embed_norm                 │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ pos_drop (Dropout)              │ (None, 4, 64, 64, 96)  │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_1         │ (None, 4, 32, 32, 192) │       305,274 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       115,389 │
│ video_swin_transformer_block    │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ identity (Identity)     │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization     │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │        40,797 │
│ video_swin_window_attention     │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense (Dense)        │ ?                      │        27,936 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout (Dropout)    │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_1 (Dense)      │ ?                      │         9,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_1 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_1   │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp (MLP)               │ ?                      │        74,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       115,389 │
│ video_swin_transformer_block_1  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path (DropPath)    │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_2   │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │        40,797 │
│ video_swin_window_attention_1   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_4 (Dense)      │ ?                      │        27,936 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_3 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_5 (Dense)      │ ?                      │         9,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_4 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_3   │ ?                      │           192 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_1 (MLP)             │ ?                      │        74,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ video_swin_patch_merging   │ ?                      │        74,496 │
│ (VideoSwinPatchMerging)         │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_2         │ (None, 4, 16, 16, 384) │     1,200,372 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       451,962 │
│ video_swin_transformer_block_2  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_1 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_5   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       155,322 │
│ video_swin_window_attention_2   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_9 (Dense)      │ ?                      │       111,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_6 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_10 (Dense)     │ ?                      │        37,056 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_7 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_6   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_2 (MLP)             │ ?                      │       295,872 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │       451,962 │
│ video_swin_transformer_block_3  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_2 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_7   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       155,322 │
│ video_swin_window_attention_3   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_13 (Dense)     │ ?                      │       111,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_9 (Dropout)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_14 (Dense)     │ ?                      │        37,056 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_10 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_8   │ ?                      │           384 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_3 (MLP)             │ ?                      │       295,872 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ video_swin_patch_merging_1 │ ?                      │       296,448 │
│ (VideoSwinPatchMerging)         │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_3         │ (None, 4, 8, 8, 768)   │    11,914,680 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_4  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_3 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_10  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_4   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_18 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_12 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_19 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_13 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_11  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_4 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_5  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_4 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_12  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_5   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_22 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_15 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_23 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_16 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_13  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_5 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_6  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_5 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_14  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_6   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_26 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_18 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_27 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_19 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_15  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_6 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_7  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_6 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_16  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_7   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_30 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_21 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_31 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_22 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_17  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_7 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_8  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_7 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_18  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_8   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_34 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_24 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_35 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_25 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_19  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_8 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     1,788,660 │
│ video_swin_transformer_block_9  │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_8 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_20  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │       605,556 │
│ video_swin_window_attention_9   │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_38 (Dense)     │ ?                      │       443,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_27 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_39 (Dense)     │ ?                      │       147,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_28 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_21  │ ?                      │           768 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_9 (MLP)             │ ?                      │     1,181,568 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └ video_swin_patch_merging_2 │ ?                      │     1,182,720 │
│ (VideoSwinPatchMerging)         │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_4         │ (None, 4, 8, 8, 768)   │    14,232,528 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     7,116,264 │
│ video_swin_transformer_block_10 │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_9 (DropPath)  │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_23  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │     2,390,760 │
│ video_swin_window_attention_10  │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_43 (Dense)     │ ?                      │     1,771,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_30 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_44 (Dense)     │ ?                      │       590,592 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_31 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_24  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_10 (MLP)            │ ?                      │     4,722,432 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│    └                            │ ?                      │     7,116,264 │
│ video_swin_transformer_block_11 │                        │               │
│ (VideoSwinTransformerBlock)     │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ drop_path_10 (DropPath) │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_25  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └                         │ ?                      │     2,390,760 │
│ video_swin_window_attention_11  │                        │               │
│ (VideoSwinWindowAttention)      │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_47 (Dense)     │ ?                      │     1,771,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_33 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dense_48 (Dense)     │ ?                      │       590,592 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│          └ dropout_34 (Dropout) │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ layer_normalization_26  │ ?                      │         1,536 │
│ (LayerNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│       └ mlp_11 (MLP)            │ ?                      │     4,722,432 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ top_norm (LayerNormalization)   │ (None, 4, 8, 8, 768)   │         1,536 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ avg_pool                        │ (None, 768)            │             0 │
│ (GlobalAveragePooling3D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ classifier (Dense)              │ (None, 400)            │       307,600 │
└─────────────────────────────────┴────────────────────────┴───────────────┘
 Total params: 27,971,494 (106.70 MB)
 Trainable params: 27,971,494 (106.70 MB)
 Non-trainable params: 0 (0.00 B)
```

With `keras.Layer` (same as `expand_nested=False`)


```python
with_layer_swin.summary(expand_nested=True)
Model: "swin_transformer3d_1"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer_1 (InputLayer)      │ (None, 8, 256, 256, 3) │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_patching_and_embeddi… │ (None, 4, 64, 64, 96)  │         9,504 │
│ (VideoSwinPatchingAndEmbedding) │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ pos_drop (Dropout)              │ (None, 4, 64, 64, 96)  │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_1         │ (None, 4, 32, 32, 192) │       305,274 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_2         │ (None, 4, 16, 16, 384) │     1,200,372 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_3         │ (None, 4, 8, 8, 768)   │    11,914,680 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ videoswin_basic_layer_4         │ (None, 4, 8, 8, 768)   │    14,232,528 │
│ (VideoSwinBasicLayer)           │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ top_norm (LayerNormalization)   │ (None, 4, 8, 8, 768)   │         1,536 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ avg_pool                        │ (None, 768)            │             0 │
│ (GlobalAveragePooling3D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ classifier (Dense)              │ (None, 400)            │       307,600 │
└─────────────────────────────────┴────────────────────────┴───────────────┘
 Total params: 27,971,494 (106.70 MB)
 Trainable params: 27,971,494 (106.70 MB)
 Non-trainable params: 0 (0.00 B)
```

Some [attempt](https://github.com/keras-team/keras/issues/15250) has been taken in past, but issue remains. Until then, is it okay to keep it as in? Additionally, I wonder why there is `?` in the output shape section?

tirthasheshpatel replied:
I agree that the summary is much more useful with layers being `keras.Model` subclasses. My suggestion was to maintain consistency across the package. That being said, I do think this might be useful if the standard suite of tests pass with these layers being `keras.Model`s.

About the output shape being `?`: Keras uses the `compute_output_shape` method of the layers to compute the output shape wherever it can. In this case, for example, the `VideoSwinBasicLayer` implements it, so Keras doesn't ever compute the shapes of the layers in the `VideoSwinBasicLayer` model.

innat replied:
> My suggestion was to maintain consistency across the package. That being said, I do think this might be useful if the standard suite of tests pass with these layers being keras.Models.

Understood. IMHO, using `keras.Layer` or `keras.Model` don't violate much on consistency. Heck, they're both same, whereas `keras.Model` brings extra stuff, like compile or fit. So, what do you prefer most? Should I update to maintain consistency or keep it as it in? youre call.

tirthasheshpatel replied:
Let's keep it this way. It seems useful enough without introducing much complexity.

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+        return config
+
+
+class VideoSwinBasicLayer(keras.Model):
```
```suggestion
class VideoSwinBasicLayer(keras.Layer):
```

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+@keras_cv_export(
+    "keras_cv.layers.VideoSwinTransformerBlock", package="keras_cv.layers"
+)
+class VideoSwinTransformerBlock(keras.Model):
```
```suggestion
class VideoSwinTransformerBlock(keras.Layer):
```

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+        return config
+
+
+@keras_cv_export(
+    "keras_cv.layers.VideoSwinTransformerBlock", package="keras_cv.layers"
+)
```
Let's also not export this if it isn't present in the public API

tirthasheshpatel commented on keras_cv/models/classification/video_classifier.py:
```diff
@@ -0,0 +1,149 @@
+        "keras_cv.models.classification.VideoClassifier",
+    ]
+)
+class VideoClassifier(Task):
```
This also needs to go in the `keras_cv.models` `__init__.py` file in order to be present in the public API.

innat replied:
this is already import in the `keras_cv.models.__init__.py` file, LOC 199, isn't it?

tirthasheshpatel replied:
Ah, missed the import, thanks!

tirthasheshpatel commented on keras_cv/layers/video_swin_layers.py:
```diff
@@ -0,0 +1,980 @@
+from keras_cv.layers import DropPath
+
+
+def window_partition(x, window_size):
```
I wonder if this and window unpartition functions can be merged with the [`keras_cv.layers.vit_det_layers.WindowPartitioning`](https://github.com/keras-team/keras-cv/blob/c123d51b5d911f94eba471ca82743973f20a6e38/keras_cv/layers/vit_det_layers.py#L302). I will leave this here as a note. No need to do it in this PR.

innat replied:
Thanks. If a component is too frequent to many models or layers, then we can seek a way to unify them. Otherwise it is better to keep them to their dedicated script. Also, there can be 2D and 3D input shape for a common layer, such as above `window partition`.  Let's explore this layer.

tirthasheshpatel commented on keras_cv/layers/__init__.py:
```diff
@@ -135,6 +135,9 @@
 )
 from keras_cv.layers.spatial_pyramid import SpatialPyramidPooling
 from keras_cv.layers.transformer_encoder import TransformerEncoder
+from keras_cv.layers.video_swin_layers import VideoSwinBasicLayer
+from keras_cv.layers.video_swin_layers import VideoSwinPatchingAndEmbedding
+from keras_cv.layers.video_swin_layers import VideoSwinPatchMerging
```
We probably don't want these in the public API, just internal layers for `VideoSwinTransformer`. What do you think? Was there a reason to make them public?

innat replied:
There is no reason to place them here. I was just keeping the consistency with other component.  For example, all of the `vit_det_layers` are placed here. Anyway, will remove it.

tirthasheshpatel replied:
It was a mistake to export the `vit_det_layers`. Other than something like `WindowPartitioning`, other layers are too specific for the detectron2 ViT model.

It would be nice to keep the layers API clean. Thanks for removing these. We can always add again if there's a usecase for it elsewhere!

innat replied:
Got it. And note, I'm removing imports that I added in this PR. Others import, like `vit_det` or anything else, let's do that later (coz some evaluation is needed of their presence here).

tirthasheshpatel reviewed (commented):
Amazing work here 🎉 Thanks @innat! I still need to test presets.

The model looks very good overall, just some nits about exporting some layers and models.

innat commented on keras_cv/models/backbones/video_swin/video_swin_aliases.py:
```diff
@@ -0,0 +1,158 @@
+        return {
+            "videoswin_base_kinetics400": copy.deepcopy(
+                backbone_presets["videoswin_base_kinetics400"]
+            ),
```
The backbone base model has more than one checkpints. 

1. with kinetics-400-base (current)
2. with kinetics-400-base-imagenet22k
3. with kinetics-600-base-imagenet22k
4. with something-something-v2

How to facilitate the `preset` method for all of these? 

```python
def presets(cls):
    """Dictionary of preset names and configurations."""
    return {
        "videoswin_base_kinetics400": copy.deepcopy(
            backbone_presets["videoswin_base_kinetics400"]
        ),
        "videoswin_base_kinetics400_imagenet22k": copy.deepcopy(
            backbone_presets["videoswin_base_kinetics400_imagenet22k"]
        ),
        ...
    }
```

innat commented:
Summarizing weight check.

**Backbones** (tolerance 1e-4)
- [notebook-1](https://www.kaggle.com/code/ipythonx/kerascv-torchvision-video-swin-logits-back)
- [notebook-2](https://www.kaggle.com/code/ipythonx/k600-ssv2-logit-matching-torch-vs-keras-cv-back)

**Classifier** (tolerance 1e-5)

- [notebook-1](https://www.kaggle.com/code/ipythonx/kerascv-torchvision-video-swin-logits-comparison)
- [notebook-2](https://www.kaggle.com/code/ipythonx/k600-ssv2-logit-matching-torch-vs-keras-cv)

notebook-1 for kinetics-400 (tiny, small, base, base-imagenet22k)
notebook-1 for kinetics-600 (base-imagenet22k), something-something-v2

innat commented:
# ONNX

I noticed others also tried to export this model to onnx format but failed and reported to the official repo, [tickets](https://github.com/SwinTransformer/Video-Swin-Transformer/issues?q=is%3Aissue+is%3Aopen+onnx). So, I tried with this implementation with torch backend and it works as expected. 

```python
model = VideoClassifier(
    backbone=backbone,
    num_classes=num_classes,
    activation=None,
    pooling='avg',
)
model.eval()
```
```python
batch_size = 1

#Input to the model
x = torch.randn(batch_size, 32, 224, 224, 3, requires_grad=True)
torch_out = model(x)
```

Using the torch official [guideline](https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html). 

```python
torch.onnx.export(
    model, # model being run
    x,  # model input (or a tuple for multiple inputs)
    "vswin.onnx", 
    export_params=True,       
    opset_version=10,       
    do_constant_folding=True, 
    input_names = ['input'],   # the model's input names
    output_names = ['output'], # the model's output names
    dynamic_axes={
        'input' : {0 : 'batch_size'}, 
        'output' : {0 : 'batch_size'}
    }
)
```
```python
import onnx
import onnxruntime

def to_numpy(tensor):
    if tensor.requires_grad:
        tensor = tensor.detach()
    tensor = tensor.cpu()
    numpy_array = tensor.numpy()
    return numpy_array

onnx_model = onnx.load("vswin.onnx")
onnx.checker.check_model(onnx_model)

ort_session = onnxruntime.InferenceSession(
    "super_resolution.onnx", providers=["CPUExecutionProvider"]
)

# compute ONNX Runtime output prediction
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
ort_outs = ort_session.run(None, ort_inputs)
```

Logit checking.

```python
np.testing.assert_allclose(
    to_numpy(torch_out), ort_outs[0], rtol=1e-03, atol=1e-05
)
```

divyashreepathihalli reviewed (changes requested):
lets move the video_swin layers into the model folder itself. Everything else LGTM!

innat commented:
> lets move the video_swin layers into the model folder itself. Everything else LGTM!

Sorry, could u please elaborate? 
Do u want [this](https://github.com/innat/keras-cv/blob/video_swin/keras_cv/layers/video_swin_layers.py) file to relocate to [here](https://github.com/innat/keras-cv/tree/video_swin/keras_cv/models/backbones/video_swin)? If so, wouldn't it be anti pattern from current standard? I mean, all of the layers supposed to be in [this directory](https://github.com/innat/keras-cv/tree/video_swin/keras_cv/layers), or no?

divyashreepathihalli commented:
> > lets move the video_swin layers into the model folder itself. Everything else LGTM!
> 
> Sorry, could u please elaborate? Do u want [this](https://github.com/innat/keras-cv/blob/video_swin/keras_cv/layers/video_swin_layers.py) file to relocate to [here](https://github.com/innat/keras-cv/tree/video_swin/keras_cv/models/backbones/video_swin)? If so, wouldn't it be anti pattern from current standard? I mean, all of the layers supposed to be in [this directory](https://github.com/innat/keras-cv/tree/video_swin/keras_cv/layers), or no?

Nope! all model specific layers should be inside the model folder. Only generic layers will go under the layers folder.

innat commented:
> Nope! all model specific layers should be inside the model folder. Only generic layers will go under the layers folder.
> The move locations linked are correct.

Could you please refer to any implementation as reference? Coz, All I can see that, layers of vit or vit_det or segformer are placed under layer directory, wheras their corresponding backbone and classifier are located in the model directory (unlesd it's a mistake.)

innat commented:
I think the test is failling for other issue.

divyashreepathihalli commented:
Thank you for this awesome contribution!!!
